## Real or Not? NLP with Disaster Tweets - Project 2 - Sam's part

Big-Scale Analytics - Project 2 - Team Rolex
> Samuel Lew, Alexandre Lang, Samy Bouzerda, Alix Muller

## Load dataset

In [3]:
import pandas as pd

In [4]:
df = pd.read_csv ('../data/train.csv')
df = df[['id', 'text', 'target']]
df.iloc[:10]

id                                               text  target
0   1  Our Deeds are the Reason of this #earthquake M...       1
1   4             Forest fire near La Ronge Sask. Canada       1
2   5  All residents asked to 'shelter in place' are ...       1
3   6  13,000 people receive #wildfires evacuation or...       1
4   7  Just got sent this photo from Ruby #Alaska as ...       1
5   8  #RockyFire Update => California Hwy. 20 closed...       1
6  10  #flood #disaster Heavy rain causes flash flood...       1
7  13  I'm on top of the hill and I can see a fire in...       1
8  14  There's an emergency evacuation happening now ...       1
9  15  I'm afraid that the tornado is coming to our a...       1

In [5]:
index = df.index
columns = df.columns
values = df.values

In [10]:
df_yes = df.query('target == 1')
yes_count = 0
for row in df_yes.iterrows():
    yes_count += 1
print(f'there are {yes_count} tweets about disasters')

df_no = df.query('target == 0')
no_count = 0
for row in df_no.iterrows():
    no_count += 1
print(f'there are {no_count} tweets that are not about disasters')

print('the classification base rate is ' + str(yes_count/(yes_count+no_count)*100)[:5] + '%')

there are 3271 tweets about disasters
there are 4342 tweets that are not about disasters
the classification base rate is 42.96%


## Process text

In [11]:
#!pip install spacy
#!python -m spacy download en

### Removing stopwords & lowercasing every word:

In [27]:
import string
import spacy
from spacy.lang.en.stop_words import STOP_WORDS
from spacy.lang.en import English

punctuations = string.punctuation

nlp = spacy.load('en')
stop_words = spacy.lang.en.stop_words.STOP_WORDS

parser = English()

filtered_tweets = []
for tweet in df['text']:
    tweet = nlp(tweet)
    filt_tweet = ' '
    for word in tweet:
        if word.is_stop == False and word.lemma_ != "-PRON-" and word not in stop_words or punctuations:
            filt_tweet = filt_tweet + ' ' + str(word)
          
    filtered_tweets.append(filt_tweet)
    
df['filtered_tweet'] = filtered_tweets

def clean_text(text):
    # Removing spaces and converting text into lowercase
    return text.strip().lower()

df['filtered_tweet'] = df['filtered_tweet'].map(clean_text)

df.head(30)

id                                               text  target  \
0    1  Our Deeds are the Reason of this #earthquake M...       1   
1    4             Forest fire near La Ronge Sask. Canada       1   
2    5  All residents asked to 'shelter in place' are ...       1   
3    6  13,000 people receive #wildfires evacuation or...       1   
4    7  Just got sent this photo from Ruby #Alaska as ...       1   
5    8  #RockyFire Update => California Hwy. 20 closed...       1   
6   10  #flood #disaster Heavy rain causes flash flood...       1   
7   13  I'm on top of the hill and I can see a fire in...       1   
8   14  There's an emergency evacuation happening now ...       1   
9   15  I'm afraid that the tornado is coming to our a...       1   
10  16        Three people died from the heat wave so far       1   
11  17  Haha South Tampa is getting flooded hah- WAIT ...       1   
12  18  #raining #flooding #Florida #TampaBay #Tampa 1...       1   
13  19            #Flood in Bago Myanmar #We arrived Bago       1   
14  20  Damage to school bus on 80 in multi car crash ...       1   
15  23                                     What's up man?       0   
16  24                                      I love fruits       0   
17  25                                   Summer is lovely       0   
18  26                                  My car is so fast       0   
19  28                       What a goooooooaaaaaal!!!!!!       0   
20  31                             this is ridiculous....       0   
21  32                                  London is cool ;)       0   
22  33                                        Love skiing       0   
23  34                              What a wonderful day!       0   
24  36                                           LOOOOOOL       0   
25  37                     No way...I can't eat that shit       0   
26  38                              Was in NYC last week!       0   
27  39                                 Love my girlfriend       0   
28  40                                          Cooool :)       0   
29  41                                 Do you like pasta?       0   

                                       filtered_tweet  
0   our deeds are the reason of this # earthquake ...  
1             forest fire near la ronge sask . canada  
2   all residents asked to ' shelter in place ' ar...  
3   13,000 people receive # wildfires evacuation o...  
4   just got sent this photo from ruby # alaska as...  
5   # rockyfire update = > california hwy . 20 clo...  
6   # flood # disaster heavy rain causes flash flo...  
7   i 'm on top of the hill and i can see a fire i...  
8   there 's an emergency evacuation happening now...  
9   i 'm afraid that the tornado is coming to our ...  
10        three people died from the heat wave so far  
11  haha south tampa is getting flooded hah- wait ...  
12  # raining # flooding # florida # tampabay # ta...  
13          # flood in bago myanmar # we arrived bago  
14  damage to school bus on 80 in multi car crash ...  
15                                   what 's up man ?  
16                                      i love fruits  
17                                   summer is lovely  
18                                  my car is so fast  
19                 what a goooooooaaaaaal ! ! ! ! ! !  
20                            this is ridiculous ....  
21                                  london is cool ;)  
22                                        love skiing  
23                             what a wonderful day !  
24                                           looooool  
25                  no way ... i ca n't eat that shit  
26                             was in nyc last week !  
27                                 love my girlfriend  
28                                          cooool :)  
29                                do you like pasta ?

In [22]:
'''for row in df:
  
  for word in ['filtered_tweet']:
    print(word.lemma)

[word.lemma_.lower().strip() for word in df['filtered_tweet']]'''

AttributeError: 'str' object has no attribute 'lemma'

### Logistic regression:

In [23]:
#!python -m spacy download en_core_web_sm

from sklearn.feature_extraction.text import CountVectorizer,TfidfVectorizer
from sklearn.base import TransformerMixin
from sklearn.pipeline import Pipeline


In [24]:
import string
from spacy.lang.en.stop_words import STOP_WORDS
from spacy.lang.en import English

def spacy_tokenizer(sentence):
    mytokens = parser(sentence)
    mytokens = [ word.lemma_.lower().strip() 
                if word.lemma_ != "-PRON-" else word.lower_ for word in mytokens ]
    mytokens = [ word for word in mytokens 
                if word not in stop_words and word not in punctuations ]

    return mytokens


tfidf_vector = TfidfVectorizer(tokenizer = spacy_tokenizer)

In [30]:
from sklearn.model_selection import train_test_split

X = df['text']
ylabels = df['target']

X_train, X_test, y_train, y_test = train_test_split(X, ylabels, test_size=0.3, random_state=72)

In [31]:
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(solver="lbfgs")

# Create pipeline using Bag of Words
pipe = Pipeline([('vectorizer', tfidf_vector),
                 ('classifier', classifier)])

# model generation
pipe.fit(X_train,y_train)

from sklearn import metrics
predicted = pipe.predict(X_test)

print(" test Accuracy:",metrics.accuracy_score(y_test, predicted))
print(" Precision:",metrics.precision_score(y_test, predicted, average=None))
print(" Recall:",metrics.recall_score(y_test, predicted, average=None))

 test Accuracy: 0.8047285464098074
 Precision: [0.78066667 0.85076531]
 Recall: [0.90916149 0.66967871]


#### we can see here that our model identified a tweet about a disaster 80.47% of the time
#### when it predicted a tweet to be about a disaster, it was correctly assessed 78.07% of the time, and was correctly assessed for a tweet not about a disaster 85.08% of the time
#### when given a tweet about a disaster, the model considered it as being about a disaster 90.92% of the time, and when given a tweet not about a disaster the model considered it as not a disaster for 66.97% of the time